src/VisualStudio/Core/Def/Implementation/UnusedReferences/Dialog/UnusedReferencesTableProvider.DataSource.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.UnusedReferences;
+using Microsoft.VisualStudio.Shell.TableManager;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.UnusedReferences.Dialog
+{
+    internal partial class UnusedReferencesTableProvider
+    {
+        internal class UnusedReferencesDataSource : ITableDataSource
+        {
+            public string SourceTypeIdentifier => nameof(UnusedReferencesDataSource);
+
+            public string Identifier => nameof(UnusedReferencesDataSource);
+
+            public string? DisplayName => null;
+
+            private ImmutableList<SinkManager> _managers = ImmutableList<SinkManager>.Empty;
+            private ImmutableArray<UnusedReferencesEntry> _currentEntries;
+
+            public IDisposable Subscribe(ITableDataSink sink)
+            {
+                return new SinkManager(this, sink);
```

Comment:
This seems to be violating the stated rule in the docs:

> If, when the call is made to subscribe to a ITableDataSource, the source already has entries, then the source needs to add these entries using sink. It can make the call to add the entries before returning from the Subscribe(ITableDataSink) call.

Now maybe the assumption here is this works because we call ReplaceData exactly once, and then notifies about the entries. If that's how this is designed to work then maybe this is fine, but that probably needs some heavy commenting.

Reply:
It also may explain why you're able to get away with not telling sinks about the items being removed. :smile: